File: app/src/test/java/com/nexaflow/saayalite/domain/engine/RulesTest.kt
package com.nexaflow.saayalite.domain.engine

import com.nexaflow.saayalite.domain.model.RiskTier
import java.time.Duration
import java.time.Instant
import java.time.LocalTime
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class RulesTest {
    private val rules = Rules.DEFAULT
    private val armingEvaluator = ArmingEvaluator(rules)
    private val intervalCalculator = IntervalCalculator(rules)

    @Test
    fun `hour band boundaries use the frozen India-local partitions`() {
        assertEquals(HourBand.NIGHT_LATE, rules.hourBand(LocalTime.MIDNIGHT.minusMinutes(1)))
        assertEquals(HourBand.NIGHT_DEEP, rules.hourBand(LocalTime.MIDNIGHT))
        assertEquals(
            HourBand.NIGHT_DEEP,
            rules.hourBand(LocalTime.of(Rules.DAWN_START_HOUR, 0).minusMinutes(1)),
        )
        assertEquals(HourBand.DAWN, rules.hourBand(LocalTime.of(Rules.DAWN_START_HOUR, 0)))
        assertEquals(
            HourBand.DAY,
            rules.hourBand(LocalTime.of(Rules.NIGHT_EARLY_START_HOUR, 0).minusMinutes(1)),
        )
        assertEquals(
            HourBand.NIGHT_EARLY,
            rules.hourBand(LocalTime.of(Rules.NIGHT_EARLY_START_HOUR, 0)),
        )
    }

    @Test
    fun `arming matrix matches all cells exactly`() {
        val armedCells =
            setOf(
                RiskTier.HIGH to HourBand.NIGHT_EARLY,
                RiskTier.HIGH to HourBand.NIGHT_LATE,
                RiskTier.HIGH to HourBand.NIGHT_DEEP,
                RiskTier.HIGH to HourBand.DAWN,
                RiskTier.ELEVATED to HourBand.NIGHT_LATE,
                RiskTier.ELEVATED to HourBand.NIGHT_DEEP,
                RiskTier.ELEVATED to HourBand.DAWN,
                RiskTier.MODERATE to HourBand.NIGHT_DEEP,
            )

        RiskTier.entries.forEach { tier ->
            HourBand.entries.forEach { band ->
                assertEquals(
                    "$tier at $band",
                    tier to band in armedCells,
                    armingEvaluator.shouldAutoArm(
                        riskTier = tier,
                        hourBand = band,
                        cooldownUntil = null,
                        now = Instant.EPOCH,
                    ),
                )
            }
        }
    }

    @Test
    fun `safe never auto arms in any hour band`() {
        assertTrue(
            HourBand.entries.none {
                armingEvaluator.shouldAutoArm(
                    RiskTier.SAFE,
                    it,
                    cooldownUntil = null,
                    now = Instant.EPOCH,
                )
            },
        )
    }

    @Test
    fun `interval selection follows tier hour and manual rules`() {
        assertEquals(
            Duration.ofMinutes(5).seconds.toInt(),
            intervalCalculator.seconds(RiskTier.HIGH, HourBand.NIGHT_DEEP, ArmMode.AUTO_ZONE),
        )
        assertEquals(
            Duration.ofMinutes(12).seconds.toInt(),
            intervalCalculator.seconds(RiskTier.MODERATE, HourBand.NIGHT_DEEP, ArmMode.AUTO_ZONE),
        )
        HourBand.entries.forEach { band ->
            assertEquals(
                Duration.ofMinutes(10).seconds.toInt(),
                intervalCalculator.seconds(null, band, ArmMode.MANUAL),
            )
        }
    }

    @Test
    fun `display risk applies multiplier labels and clamps`() {
        val risk = rules.displayRisk(0.5, HourBand.NIGHT_DEEP)
        assertEquals(0.65, risk, 0.0)
        assertEquals(DisplayRiskLabel.ELEVATED, rules.displayRiskLabel(risk))

        val clamped = rules.displayRisk(Double.MAX_VALUE, HourBand.NIGHT_DEEP)
        assertEquals(1.0, clamped, 0.0)
        assertEquals(DisplayRiskLabel.HIGH, rules.displayRiskLabel(clamped))
    }

    @Test
    fun `weak PIN rules reject frozen examples only as specified`() {
        assertTrue(rules.isWeakPin("0000"))
        assertTrue(rules.isWeakPin("1234"))
        assertTrue(rules.isWeakPin("1111"))
        assertTrue(rules.isWeakPin("7777"))
        assertFalse(rules.isWeakPin("4062"))
    }
}

File: src/domain/engine/rules.test.ts
import { describe, expect, it } from "vitest";

import { shouldAutoArm } from "./armingEvaluator";
import {
  DEFAULT_RULES,
  DAWN_START_HOUR,
  DEMO_ARM_HOUR,
  DEMO_ARM_TIME,
  DEMO_DIVISOR,
  DEMO_RULES,
  HOURS_PER_DAY,
  MINUTES_PER_HOUR,
  NIGHT_DEEP_START_HOUR,
  NIGHT_EARLY_START_HOUR,
  NIGHT_LATE_START_HOUR,
  NORMAL_DEMO_DIVISOR,
  PIN_LENGTH,
  WESTERN_DIGIT_MIN,
  displayRisk,
  displayRiskLabel,
  hourBandForLocalTime,
  isWeakPin,
  scaledSeconds,
} from "./rules";
import type { HourBand, RiskTier } from "../model/session";

const SECONDS_PER_MINUTE = 60; // GROUNDED-EXEMPT: SI unit conversion in test expectations
const ALL_BANDS: readonly HourBand[] = [
  "DAY",
  "NIGHT_EARLY",
  "NIGHT_LATE",
  "NIGHT_DEEP",
  "DAWN",
];

const EXPECTED_ARMING: Readonly<Record<RiskTier, readonly HourBand[]>> = {
  HIGH: ["NIGHT_EARLY", "NIGHT_LATE", "NIGHT_DEEP", "DAWN"],
  ELEVATED: ["NIGHT_LATE", "NIGHT_DEEP", "DAWN"],
  MODERATE: ["NIGHT_DEEP"],
  SAFE: [],
};

describe("frozen business rules", () => {
  it("resolves every inclusive hour-band boundary", () => {
    const lastMinute = MINUTES_PER_HOUR - 1;
    expect(hourBandForLocalTime(HOURS_PER_DAY - 1, lastMinute)).toBe(
      "NIGHT_LATE",
    );
    expect(hourBandForLocalTime(NIGHT_DEEP_START_HOUR, 0)).toBe("NIGHT_DEEP");
    expect(hourBandForLocalTime(DAWN_START_HOUR - 1, lastMinute)).toBe(
      "NIGHT_DEEP",
    );
    expect(hourBandForLocalTime(DAWN_START_HOUR, 0)).toBe("DAWN");
    expect(hourBandForLocalTime(NIGHT_EARLY_START_HOUR - 1, lastMinute)).toBe(
      "DAY",
    );
    expect(hourBandForLocalTime(NIGHT_EARLY_START_HOUR, 0)).toBe(
      "NIGHT_EARLY",
    );
    expect(hourBandForLocalTime(NIGHT_LATE_START_HOUR, 0)).toBe("NIGHT_LATE");
  });

  it("matches every arming-matrix cell", () => {
    (Object.keys(EXPECTED_ARMING) as RiskTier[]).forEach((tier) => {
      ALL_BANDS.forEach((band) => {
        expect(shouldAutoArm(DEFAULT_RULES, tier, band, undefined, 0)).toBe(
          EXPECTED_ARMING[tier].includes(band),
        );
      });
    });
  });

  it("keeps DAY non-arming and derives the demo band from its frozen hour", () => {
    (Object.keys(EXPECTED_ARMING) as RiskTier[]).forEach((tier) => {
      expect(shouldAutoArm(DEFAULT_RULES, tier, "DAY", undefined, 0)).toBe(
        false,
      );
    });
    expect(DEMO_ARM_HOUR).toBe(4);
    expect(DEMO_ARM_TIME).toEqual({
      hourBand: hourBandForLocalTime(DEMO_ARM_HOUR, 0),
      hourOfDay: DEMO_ARM_HOUR,
    });
    expect(DEMO_ARM_TIME.hourBand).toBe("NIGHT_DEEP");
  });

  it("never auto-arms a safe zone", () => {
    ALL_BANDS.forEach((band) => {
      expect(shouldAutoArm(DEFAULT_RULES, "SAFE", band, undefined, 0)).toBe(
        false,
      );
    });
  });

  it("freezes the normal three-check-in ladder timing profile", () => {
    expect(DEFAULT_RULES.ladder).toEqual({
      cadenceSec: 5 * MINUTES_PER_HOUR,
      window1Sec: 2 * MINUTES_PER_HOUR,
      window2Sec: 1 * MINUTES_PER_HOUR,
      window3Sec: SECONDS_PER_MINUTE,
      okResetSec: 5 * MINUTES_PER_HOUR,
    });
  });

  it("compresses only the demo windows and OK reset, keeping the cadence", () => {
    expect(DEMO_RULES.ladder).toEqual({
      cadenceSec: 5 * MINUTES_PER_HOUR,
      window1Sec: 10,
      window2Sec: 10,
      window3Sec: 10,
      okResetSec: 10,
    });
  });

  it("runs both profiles unscaled since the divisor was retired", () => {
    expect(DEMO_DIVISOR).toBe(1);
    expect(NORMAL_DEMO_DIVISOR).toBe(1);
    expect(DEFAULT_RULES.demoDivisor).toBe(1);
    expect(DEMO_RULES.demoDivisor).toBe(1);
    expect(scaledSeconds(SECONDS_PER_MINUTE, DEMO_RULES)).toBe(
      SECONDS_PER_MINUTE,
    );
    expect(scaledSeconds(SECONDS_PER_MINUTE, DEFAULT_RULES)).toBe(
      SECONDS_PER_MINUTE,
    );
  });

  it("modulates, labels and clamps display risk", () => {
    const nightRisk = displayRisk(0.5, "NIGHT_DEEP");
    expect(nightRisk).toBe(0.65);
    expect(displayRiskLabel(nightRisk)).toBe("Elevated");
    expect(displayRisk(1, "NIGHT_DEEP")).toBe(1);
  });

  it("rejects every frozen weak PIN example", () => {
    const repeatedZeroPin = WESTERN_DIGIT_MIN.repeat(PIN_LENGTH);
    expect(isWeakPin(repeatedZeroPin)).toBe(true);
    expect(isWeakPin("1234")).toBe(true);
    expect(isWeakPin("1111")).toBe(true);
    expect(isWeakPin("7777")).toBe(true);
    expect(isWeakPin("4062")).toBe(false);
  });
});